Server/ReasnAPI/ReasnAPI/Controllers/UsersController.cs

```diff
@@ -1,47 +1,91 @@
+using FluentValidation;
 using Microsoft.AspNetCore.Authorization;
 using Microsoft.AspNetCore.Mvc;
+using ReasnAPI.Models.DTOs;
+using ReasnAPI.Models.Enums;
+using ReasnAPI.Services;
 
 namespace ReasnAPI.Controllers;
 
 [ApiController]
 [Route("[controller]")]
-public class UsersController : ControllerBase
+public class UsersController(UserService userService, InterestService interestService) : ControllerBase
 {
+    private readonly UserService _userService = userService;
+    private readonly InterestService _interestService = interestService;
+
     [HttpGet]
     [Authorize(Roles = "Admin")]
+    [ProducesResponseType<IEnumerable<UserDto>>(StatusCodes.Status200OK)]
     public IActionResult GetUsers()
     {
-        throw new NotImplementedException();
+        var users = _userService.GetUsersByFilter(u => u.IsActive && u.Role != UserRole.Admin);
+        return Ok(users);
     }
 
     [HttpGet]
     [Route("{username}")]
-    public IActionResult GetUserByUsername(string username)
+    [ProducesResponseType<UserDto>(StatusCodes.Status200OK)]
+    public IActionResult GetUserByUsername([FromRoute] string username)
     {
-        throw new NotImplementedException();
+        var user = _userService.GetUserByUsername(username);
+
+        if (user.Role == UserRole.Admin)
+        {
+            return Forbid();
+        }
+
+        return Ok(user);
     }
 
     [HttpPut]
     [Authorize]
     [Route("{username}")]
-    public IActionResult UpdateUser(string username)
+    [ProducesResponseType<UserDto>(StatusCodes.Status200OK)]
+    public IActionResult UpdateUser(
+        [FromBody] UserDto userDto,
+        [FromRoute] string username,
+        [FromServices] IValidator<UserDto> validator)
     {
-        throw new NotImplementedException();
-    }
+        validator.ValidateAndThrow(userDto);
 
-    [HttpGet]
-    [Authorize]
-    [Route("interests")]
-    public IActionResult GetUsersInterests(string username)
-    {
-        throw new NotImplementedException();
-    }
+        var currentUser = _userService.GetCurrentUser();
+        var userToUpdate = _userService.GetUserByUsername(username);
 
-    [HttpDelete]
-    [Authorize(Roles = "Admin")]
-    [Route("interests/{interestId:int}")]
-    public IActionResult DeleteUserInterest(int interestId)
+        // Users can only update their own profile, unless they are an admin
+        if (currentUser.Role != UserRole.Admin && currentUser.Username != userToUpdate.Username)
+        {
+            return Forbid();
+        }
```

Comment:
We can do it propably before getting userToUpdate, just preventing unnecessary calls to DB